pkg/controller/endpointslicemirroring/reconciler.go

```diff
@@ -0,0 +1,405 @@
+// EndpointSlices exist. It creates, updates, or deletes EndpointSlices to
+// ensure the desired set of addresses are represented by EndpointSlices.
+func (r *reconciler) reconcile(endpoints *corev1.Endpoints, existingSlices []*discovery.EndpointSlice) error {
+	slicesToCreate := []*discovery.EndpointSlice{}
+	slicesToUpdate := []*discovery.EndpointSlice{}
+	slicesToDelete := []*discovery.EndpointSlice{}
+
+	// Build data structures for existing state.
+	existingSlicesByKey := map[addrTypePortMapKey][]*discovery.EndpointSlice{}
+	numExistingEndpoints := 0
+	for _, existingSlice := range existingSlices {
+		epHash := newAddrTypePortMapKey(existingSlice.Ports, existingSlice.AddressType)
+		existingSlicesByKey[epHash] = append(existingSlicesByKey[epHash], existingSlice)
+		numExistingEndpoints += len(existingSlice.Endpoints)
+	}
+
+	// Build data structures for desired state.
+	desiredPortsByKey := map[addrTypePortMapKey][]discovery.EndpointPort{}
+	desiredEndpointsByKey := map[addrTypePortMapKey]endpointSet{}
+	numDesiredEndpoints := 0
+
+	for _, subset := range endpoints.Subsets {
+		endpointPorts := epPortsToEpsPorts(subset.Ports)
+		v4Hash := newAddrTypePortMapKey(endpointPorts, discovery.AddressTypeIPv4)
+		v6Hash := newAddrTypePortMapKey(endpointPorts, discovery.AddressTypeIPv6)
+		if _, ok := desiredEndpointsByKey[v4Hash]; !ok {
+			desiredEndpointsByKey[v4Hash] = endpointSet{}
+		}
+		if _, ok := desiredEndpointsByKey[v6Hash]; !ok {
+			desiredEndpointsByKey[v6Hash] = endpointSet{}
+		}
+
+		desiredPortsByKey[v4Hash] = endpointPorts
+		desiredPortsByKey[v6Hash] = endpointPorts
+		totalAddresses := 0
+
+		for _, address := range subset.Addresses {
+			totalAddresses++
+			if totalAddresses > maxEndpointsPerSubset {
+				break
+			}
+			endpoint := addressToEndpoint(address, true)
+			addrType := getAddressType(address.IP)
+			if addrType == nil {
+				klog.Warningf("Address in %s/%s Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice: %s", endpoints.Namespace, endpoints.Name, address.IP)
+			} else {
+				epHash := v4Hash
+				if *addrType == discovery.AddressTypeIPv6 {
+					epHash = v6Hash
+				}
+				desiredEndpointsByKey[epHash].Insert(endpoint)
+				numDesiredEndpoints++
+			}
+		}
+
+		for _, address := range subset.NotReadyAddresses {
+			totalAddresses++
+			if totalAddresses > maxEndpointsPerSubset {
+				break
+			}
+			endpoint := addressToEndpoint(address, false)
+			addrType := getAddressType(address.IP)
+			if addrType == nil {
+				klog.Warningf("Address in %s/%s Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice: %s", endpoints.Namespace, endpoints.Name, address.IP)
+			} else {
+				epHash := v4Hash
+				if *addrType == discovery.AddressTypeIPv6 {
+					epHash = v6Hash
+				}
+				desiredEndpointsByKey[epHash].Insert(endpoint)
+				numDesiredEndpoints++
+			}
+		}
+	}
+
+	epMetrics := mirroringmetrics.NewEndpointPortCache()
+	totalAdded := 0
```

Comment:
total := struct { added, removed int }

Reply:
Good idea, updated all uses of that pattern to use this kind of struct.